mgr/nfs: NFS cluster and export commands to enable and disable ops control

ShwetaBhosale1 opened the pull request:
mgr/nfs: NFS cluster and export commands to enable and disable ops/s control

Fixes: https://tracker.ceph.com/issues/69861
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

For bandwidth control description, check https://github.com/ceph/ceph/pull/61283

**Cluster commands:**
1. _nfs cluster qos enable ops_control <cluster_id> <qos_type:PerShare|PerClient|PerShare_PerClient> [<max_export_iops:int>] [<max_client_iops:int>]_
2. _nfs cluster qos disable ops_control <cluster_id>_

For enable command:
a. If qos_type is pershare, then max_export_iops is  compulsory
b. If qos_type is perclient, then max_client_iops is compulsory
c. If qos_type is pershare_perclient then both max_export_iops and max_client_iops are compulsory parameters

**Export commands:**
1. _nfs export qos enable ops_control <cluster_id> <pseudo_path> [<max_export_iops:int>] [<max_client_iops:int>]_
2. _nfs export qos disable ops_control <cluster_id> <pseudo_path>_

For enable command:
a. If qos_type is pershare, then max_export_iops is  compulsory
b. If qos_type is perclient, then ops control enable is not allowed at export level.
c. If qos_type is pershare_perclient then both max_export_iops and max_client_iops are compulsory parameters

Qos_type is the common parameter of both bandwidths and ops control the command. So if it is set, then the enable command will fail if different qos_type is passed.

Testing logs:
```
[ceph: root@ceph-node-0 /]# ceph nfs cluster qos get mynfs
{}
[ceph: root@ceph-node-0 /]#
[ceph: root@ceph-node-0 /]# ceph nfs cluster qos enable bandwidth_control mynfs PerShare --max_export_write_bw 10MB --max_export_read_bw 20MB
[ceph: root@ceph-node-0 /]# ceph nfs cluster qos enable ops_control mynfs PerClient
Error EINVAL: Bandwidth control is using PerShare qos type, please update that qos type for Bandwidth first.
[ceph: root@ceph-node-0 /]# ceph nfs cluster qos enable ops_control mynfs PerShare --max_export_iops 10000
[ceph: root@ceph-node-0 /]# ceph nfs cluster qos get mynfs
{
  "combined_rw_bw_control": false,
  "enable_bw_control": true,
  "enable_ops_control": true,
  "enable_qos": true,
  "max_export_iops": 10000,
  "max_export_read_bw": "20.0MB",
  "max_export_write_bw": "10.0MB",
  "qos_type": "PerShare"
}
[ceph: root@ceph-node-0 /]# 


[ceph: root@ceph-node-0 /]# ceph nfs export qos get mynfs /export1
{}
[ceph: root@ceph-node-0 /]# ceph nfs export qos enable ops_control mynfs /export1 --max_export_iops 20000
[ceph: root@ceph-node-0 /]# ceph nfs export qos get mynfs /export1
{
  "enable_ops_control": true,
  "enable_qos": true,
  "max_export_iops": 20000
}
[ceph: root@ceph-node-0 /]# 

[ceph: root@ceph-node-0 /]# ceph nfs cluster qos disable ops_control mynfs                                  
[ceph: root@ceph-node-0 /]# ceph nfs export qos enable ops_control mynfs /export1 --max_export_iops 20000
Error EINVAL: To configure IOPS control for export, you must first enable IOPS control at the cluster level mynfs.
[ceph: root@ceph-node-0 /]# 
[ceph: root@ceph-node-0 /]# ceph nfs cluster qos enable ops_control mynfs PerShare --max_export_iops 10000
[ceph: root@ceph-node-0 /]# ceph nfs export qos enable bandwidth_control mynfs /export1  --max_export_write_bw 10MB --max_export_read_bw 20MB
[ceph: root@ceph-node-0 /]# ceph nfs export qos get mynfs /export1
{
  "combined_rw_bw_control": false,
  "enable_bw_control": true,
  "enable_ops_control": true,
  "enable_qos": true,
  "max_export_iops": 20000,
  "max_export_read_bw": "20.0MB",
  "max_export_write_bw": "10.0MB"
}
[ceph: root@ceph-node-0 /]# 
```
```
[ceph: root@ceph-node-0 /]# cat export2.json 
{
  "access_type": "RW",
  "clients": [],
  "cluster_id": "mynfs",
  "export_id": 2,
  "fsal": {
    "cmount_path": "/",
    "fs_name": "myfs",
    "name": "CEPH",
    "user_id": "nfs.mynfs.myfs.126d1904"
  },
  "path": "/",
  "protocols": [
    3,
    4
  ],
  "pseudo": "/export2",
  "security_label": true,
  "squash": "none",
  "transports": [
    "TCP"
  ],
  "qos_block": {
    "combined_rw_bw_control": false,
    "enable_bw_control": true,
    "enable_ops_control": true,
    "enable_qos": true,
    "max_export_iops": 30000,
    "max_export_read_bw": "30.0MB",
    "max_export_write_bw": "30.0MB"
  }
}
[ceph: root@ceph-node-0 /]# ceph nfs export apply mynfs -i export2.json 
[
  {
    "pseudo": "/export2",
    "state": "updated"
  }
]
[ceph: root@ceph-node-0 /]# 

```
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

ShwetaBhosale1 committed:
mgr/nfs: NFS commands to enable, disable and get QOS config for cluster and export

fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/nfs: NFS cluster and export commands to enable and disable ops/s control

Fixes: https://tracker.ceph.com/issues/69861
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 commented:
jenkins test make check

ShwetaBhosale1 commented:
jenkins test api

ShwetaBhosale1 committed:
mgr/nfs: Updating QoS min and max limit of bandwidth and iops values

fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/nfs: Updated get qos to return bandwidth in bytes as dashboard needs bw in bytes

Fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/nfs: Add cluster level qos status to export get qos command

fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 commented:
jenkins test signed

adk3798 commented:
@ShwetaBhosale1 the signed-off-by check is failing because you're missing a signed-off-by line on the commit to return the size in bytes for the dashboard

ShwetaBhosale1 commented:
jenkins test api

ShwetaBhosale1 commented:
jenkins test make check

ShwetaBhosale1 commented:
jenkins test make check arm64

ShwetaBhosale1 committed:
mgr/nfs: When cluster level qos is disabled and export still has qos parameters, then allow nfs export apply command if file has same qos block which is already set

fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 commented:
Resolved conflicts of src/pybind/mgr/nfs/module.py file.

ShwetaBhosale1 committed:
mgr/nfs: Added skip_notify_nfs_server parameter to update export commands

fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
nfs/mgr: Updated logs and exception messages

fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 commented:
jenkins test make check

phlogistonjohn commented on src/pybind/mgr/nfs/qos_conf.py:
```diff
@@ -0,0 +1,354 @@
+from ceph.utils import bytes_to_human, with_units_to_int
+
+
+class RawBlock():
```
I don't understand why you're moving a basic class like this into a more specific sounding file. Originally, I was going to complain about the constructor arguments until I realized this is a move. I don't see the point of "stealing" this class from the generic gansha conf file. If you don't think ganesha conf is general enough, you should probably create a new file (may be call it `ganesha_raw_conf.py` or something).

ShwetaBhosale1 replied:
It was causing a circular import issue, so I moved it. But creating a new file makes more sense. Thanks for pointing that out!!

phlogistonjohn commented on src/pybind/mgr/nfs/qos_conf.py:
```diff
@@ -0,0 +1,354 @@
+}
+
+
+class QOSBandwidthControl(object):
+    def __init__(self,
+                 enable_bw_ctrl: bool = False,
+                 combined_bw_ctrl: bool = False,
+                 export_writebw: str = '0',
+                 export_readbw: str = '0',
+                 client_writebw: str = '0',
+                 client_readbw: str = '0',
+                 export_rw_bw: str = '0',
+                 client_rw_bw: str = '0'
+                 ) -> None:
+        self.enable_bw_ctrl = enable_bw_ctrl
```
Please consider using the `black` formatting style for net new code.

ShwetaBhosale1 committed:
mgr/nfs: Addressd review comments

Fixes:  https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/nfs: Changed IOPS max limit to 409600 and Bandwidth max limit to 100GB

Fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/nfs: Changes for get QoS command to return bandwidth in GiB instead of GB

Fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 commented:
jenkins retest this please

ShwetaBhosale1 commented:
jenkins retest this please

ShwetaBhosale1 commented:
jenkins test api

ShwetaBhosale1 committed:
mgr/nfs: update QOS minimum BandWidth value and IOPS max value

Fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/cephadm: NFSSpec changes to apply cluster level QOS while nfs service creation

Fixes: https://tracker.ceph.com/issues/69458
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/cephadm: support nfs cluster level qos
Added below CEPH_NODES_LIST block in ganesha.conf and enable_cluster_qos in cluster level QoS block
CEPH_NODES_LIST {
        Ceph_Nodes = 192.168.100.100, 192.168.100.101, 192.168.100.102;
}

Fixes: https://tracker.ceph.com/issues/69861
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

mgr/cephadm: Changes to add NFS cluster qos inter node communication port in spec

mgr/nfs: Addressed review comments for cluster level qos support

mgr/nfs: add enable_cluster_qos = true while enabling qos

ShwetaBhosale1 commented:
Testing logs for commit [mgr/cephadm: support nfs cluster level qos](https://github.com/ceph/ceph/pull/61826/commits/01fb84724788d347887bc4e85a7a1c57c686c365)

```
[ceph: root@ceph-node-0 /]# 
[ceph: root@ceph-node-0 /]# ceph nfs cluster create mynfs '*'
[ceph: root@ceph-node-0 /]# 
[ceph: root@ceph-node-0 /]# ceph orch ls
NAME                       PORTS   RUNNING  REFRESHED  AGE  PLACEMENT  
crash                                  3/3  5m ago     24m  *          
mgr                                    2/2  5m ago     24m  count:2    
mon                                    3/5  5m ago     24m  count:5    
nfs.mynfs                  ?:2049      3/3  5m ago     5m   *          
osd.all-available-devices                9  5m ago     23m  *          
[ceph: root@ceph-node-0 /]# 
[ceph: root@ceph-node-0 /]# ceph orch ls
NAME                       PORTS   RUNNING  REFRESHED  AGE  PLACEMENT  
crash                                  3/3  7m ago     26m  *          
mgr                                    2/2  7m ago     26m  count:2    
mon                                    3/5  7m ago     26m  count:5    
nfs.mynfs                  ?:2049      3/3  7m ago     8m   *          
osd.all-available-devices                9  7m ago     26m  *          
[ceph: root@ceph-node-0 /]# 
[ceph: root@ceph-node-0 /]# rados -N mynfs -p '.nfs' ls
grace
rec-0000000000000004:2
rec-0000000000000004:1
conf-nfs.mynfs
rec-0000000000000004:0
[ceph: root@ceph-node-0 /]# 
[ceph: root@ceph-node-0 /]#  ceph nfs cluster qos enable bandwidth_control mynfs --qos_type PerShare --max_export_write_bw 100MB --max_export_read_bw 100MB
[ceph: root@ceph-node-0 /]# rados -N mynfs -p '.nfs' ls
grace
rec-0000000000000005:2
rec-0000000000000005:1
rec-0000000000000005:0
qosconf-nfs.mynfs
rec-0000000000000004:2
rec-0000000000000004:1
conf-nfs.mynfs
rec-0000000000000004:0
[ceph: root@ceph-node-0 /]# rados -N mynfs -p '.nfs' get qosconf-nfs.mynfs /qosconf-nfs.mynfs
[ceph: root@ceph-node-0 /]# cat /qosconf-nfs.mynfs 
QOS_DEFAULT_CONFIG {
    enable_qos = true;
    qos_type = 1;
    enable_bw_control = true;
    combined_rw_bw_control = false;
    max_export_write_bw = 100000000;
    max_export_read_bw = 100000000;
}
[ceph: root@ceph-node-0 /]# 
[ceph: root@ceph-node-0 /]# ceph nfs cluster qos get mynfs
{
  "combined_rw_bw_control": false,
  "enable_bw_control": true,
  "enable_iops_control": false,
  "enable_qos": true,
  "max_export_read_bw": "95.4MiB",
  "max_export_write_bw": "95.4MiB",
  "qos_type": "PerShare"
}
[ceph: root@ceph-node-0 /]# ceph nfs cluster qos enable ops_control mynfs PerShare --max_export_iops 1000 
[ceph: root@ceph-node-0 /]# rados -N mynfs -p '.nfs' get qosconf-nfs.mynfs /qosconf-nfs.mynfs
[ceph: root@ceph-node-0 /]# cat /qosconf-nfs.mynfs 
QOS_DEFAULT_CONFIG {
    enable_qos = true;
    qos_type = 1;
    enable_bw_control = true;
    combined_rw_bw_control = false;
    max_export_write_bw = 100000000;
    max_export_read_bw = 100000000;
    enable_iops_control = true;
    max_export_iops = 1000;
}
[ceph: root@ceph-node-0 /]# 
[ceph: root@ceph-node-0 /]# ceph nfs cluster cluster_qos mynfs enable --msg_interval 100ms
[ceph: root@ceph-node-0 /]# rados -N mynfs -p '.nfs' get qosconf-nfs.mynfs /qosconf-nfs.mynfs
[ceph: root@ceph-node-0 /]# cat /qosconf-nfs.mynfs 
QOS_DEFAULT_CONFIG {
    enable_qos = true;
    qos_type = 1;
    enable_cluster_qos = true;
    cqos_msg_interval = 100;
    enable_bw_control = true;
    combined_rw_bw_control = false;
    max_export_write_bw = 100000000;
    max_export_read_bw = 100000000;
    enable_iops_control = true;
    max_export_iops = 1000;
}
[ceph: root@ceph-node-0 /]# ceph nfs cluster cluster_qos mynfs enable --msg_interval 200ms
[ceph: root@ceph-node-0 /]# rados -N mynfs -p '.nfs' get qosconf-nfs.mynfs /qosconf-nfs.mynfs
[ceph: root@ceph-node-0 /]# cat /qosconf-nfs.mynfs 
QOS_DEFAULT_CONFIG {
    enable_qos = true;
    qos_type = 1;
    enable_cluster_qos = true;
    cqos_msg_interval = 200;
    enable_bw_control = true;
    combined_rw_bw_control = false;
    max_export_write_bw = 100000000;
    max_export_read_bw = 100000000;
    enable_iops_control = true;
    max_export_iops = 1000;
}
[ceph: root@ceph-node-0 /]# 
[ceph: root@ceph-node-0 /]# 
[ceph: root@ceph-node-0 /]# ceph nfs cluster cluster_qos mynfs disable                    
[ceph: root@ceph-node-0 /]# rados -N mynfs -p '.nfs' get qosconf-nfs.mynfs /qosconf-nfs.mynfs
[ceph: root@ceph-node-0 /]# cat /qosconf-nfs.mynfs 
QOS_DEFAULT_CONFIG {
    enable_qos = true;
    qos_type = 1;
    enable_cluster_qos = false;
    enable_bw_control = true;
    combined_rw_bw_control = false;
    max_export_write_bw = 100000000;
    max_export_read_bw = 100000000;
    enable_iops_control = true;
    max_export_iops = 1000;
}
[ceph: root@ceph-node-0 /]# ceph nfs cluster cluster_qos mynfs enable --msg_interval 200ms
[ceph: root@ceph-node-0 /]# rados -N mynfs -p '.nfs' get qosconf-nfs.mynfs /qosconf-nfs.mynfs
[ceph: root@ceph-node-0 /]# cat /qosconf-nfs.mynfs 
QOS_DEFAULT_CONFIG {
    enable_qos = true;
    qos_type = 1;
    enable_cluster_qos = true;
    cqos_msg_interval = 200;
    enable_bw_control = true;
    combined_rw_bw_control = false;
    max_export_write_bw = 100000000;
    max_export_read_bw = 100000000;
    enable_iops_control = true;
    max_export_iops = 1000;
}
[ceph: root@ceph-node-0 /]# ceph nfs cluster qos disable bandwidth_control mynfs 
[ceph: root@ceph-node-0 /]# rados -N mynfs -p '.nfs' get qosconf-nfs.mynfs /qosconf-nfs.mynfs
[ceph: root@ceph-node-0 /]# cat /qosconf-nfs.mynfs 
QOS_DEFAULT_CONFIG {
    enable_qos = true;
    qos_type = 1;
    enable_cluster_qos = true;
    cqos_msg_interval = 200;
    enable_bw_control = false;
    combined_rw_bw_control = false;
    enable_iops_control = true;
    max_export_iops = 1000;
}
[ceph: root@ceph-node-0 /]# ceph nfs cluster qos disable ops_control mynfs 
[ceph: root@ceph-node-0 /]# rados -N mynfs -p '.nfs' get qosconf-nfs.mynfs /qosconf-nfs.mynfs
[ceph: root@ceph-node-0 /]# cat /qosconf-nfs.mynfs 
QOS_DEFAULT_CONFIG {
    enable_qos = false;
    enable_bw_control = false;
    combined_rw_bw_control = false;
    enable_iops_control = false;
}
[ceph: root@ceph-node-0 /]# 


```

ShwetaBhosale1 commented:
jenkins test make check

ShwetaBhosale1 commented:
jenkins test docs

rkachach commented on src/pybind/mgr/nfs/qos_conf.py:
```diff
@@ -58,6 +60,19 @@ def _validate_qos_ops(count: int) -> int:
+        return 0
+    if msg_interval < min_interval or msg_interval > max_interval:
+        raise Exception(
+            f'Provided message interval is not in range, Please enter a value between {min_interval}ms and {max_interval}ms.'
```
would be good to show the value in the msg as well, it will give more info/context to the reported error.

rkachach commented on src/pybind/mgr/nfs/cluster.py:
```diff
@@ -612,15 +633,62 @@ def disable_cluster_qos_ops(self, cluster_id: str) -> None:
+                log.error(f'No existing QoS configuration found for cluster {cluster_id}. '
+                          f'Can not {action} cluster-qos')
+                raise Exception(f'No existing QoS configuration found for cluster {cluster_id}. '
+                                f'Can not {action} cluster-qos')
```
same msg, use a variable to store it to avoid dup

ShwetaBhosale1 commented on src/pybind/mgr/nfs/qos_conf.py:
```diff
@@ -58,6 +60,19 @@ def _validate_qos_ops(count: int) -> int:
+    max_interval = 300  # ms
+
+    if not msg_interval:
+        return 0
```
msg_interval 0 is treated as its not set, added comment mentioning same

rkachach reviewed (approved):
LGTM

ShwetaBhosale1 committed:
mgr/nfs: Update cluster qos_type from int to str

Fixes: https://tracker.ceph.com/issues/69861
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/nfs: 1. Removed the option to enable and disable cluster wide qos, it will be enabled by default
2. Removed the cluster_enable_qos field from the cluster-level block as it was causing confusion for the user.
3. Instead of using cluster use global while showing cluster level qos values in export qos get

Fixes: https://tracker.ceph.com/issues/69861
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/cephadm: For colocating nfs daeoms, check daemon qos ports instead of fetching port from spec

Fixes: https://tracker.ceph.com/issues/69861
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/cephadm: Fixed scheduling test for nfs qos port

Fixes: https://tracker.ceph.com/issues/69861
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

ShwetaBhosale1 committed:
mgr/cephadm: Handle daemon colocation port for cluster qos port

Fixes: https://tracker.ceph.com/issues/69861
Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

rkachach commented:
Teuthology run was successful. Failing Test cases are not related to this PR: https://pulpito.ceph.com/rkachach-2026-04-29_15:06:59-orch:cephadm-wip-rkachach-testing-2-2026-04-29-1255-distro-default-trial/
